### ISSUE TITLE
Backward problem

### ISSUE THREAD

**Issue by FeiiYin** [yuval-alaluf/hyperstyle#40]
Hi, it's a problem about the loss backward implementation.
I see in the coach, that the training process is written as follows.
```
for _iter in range(n_iter):
          forward()
          loss = cal_loss()
          loss.backward()
opt.zero_grad()
opt.step()
```
Why is the process implemented as multi backwards and only one optimizer step, e.g. for time-saving?
Can we write as one loss backwards and one optimizer step in the for loop?
```
for _iter in range(n_iter):
          forward()
          loss = cal_loss()
          loss.backward()
          opt.zero_grad()
          opt.step()
```
Another problem is in one for loop, does the former loss backward's gradient will be saved correctly and does it affect the follow backward's gradient?
Thanks in advance!
Thanks for providing such an impressive work!

**Comment by FeiiYin** [comment]
Another question is if I perform an extra loss.backward(), does the ordinary affects the result.

**Comment by yuval-alaluf** [comment]
The multiple backward calls allow us to accumulate the gradients across all the iterative steps and then perform the `optimizer.step()` call once for all the steps together.

I do believe that doing so does lead to some improvement in terms of the training time since we only need to take a single optimization step every 5 passes rather than every pass (although I'll admit I'm not an expert on this).
In addition, when training, we want to perform the 5 iterative passes and loss computations using the **same** model parameters. Therefore, we'll want to wait to call the `optimizer.step()` only after all iterative passes have been completed.

> Can we write as one loss backwards and one optimizer step in the for loop?

I believe that if you perform a `loss.backward()` followed by a `optimizer.step()` call for each step, you could get similar results, but I have never tried this myself, so am not sure what behavior you'll potentially get in doing so.

> does the former loss backward's gradient will be saved correctly and does it affect the follow backward's gradient?

Yes, if you perform multiple `loss.backward()` calls, we will accumulate the gradients across all the steps.

> Another question is if I perform an extra loss.backward(), does the ordinary affects the result.

I am not sure I understood the question. Could you please clarify?

**Comment by FeiiYin** [comment]
Thanks for your instant reply!
> Another question is if I perform an extra loss.backward(), does the ordinary affects the result.
For example, I wondered does the below two pieces of codes will get different results?

```
loss = id_loss()
loss.backward()
loss = L2_loss()
loss.backward()
opt.zero_grad()
opt.step()
```
```
loss = L2_loss()
loss.backward()
loss = id_loss()
loss.backward()
opt.zero_grad()
opt.step()
```
Thanks again!

**Comment by yuval-alaluf** [comment]
I tried a simple dummy example to check and it seems like you'll get the same results, which makes sense since we accumulate the gradients between the multiple backward calls.
Here is the example: 
```python

import torch

# Our "model"
x = torch.tensor([1., 2.], requires_grad=True)
# x_clone = x.clone().detach().requires_grad_(True)
y = 100 * x
z = 50 * x

# Compute loss
loss_1 = y.sum()
loss_2 = z.sum()

# Compute gradients of the parameters w.r.t. the loss
print(x.grad)     # None
loss_1.backward(retain_graph=True)
print(x.grad)     # tensor([100., 100.])
loss_2.backward()
print(x.grad)     # tensor([150., 150.])

# MOdify the parameters by subtracting the gradient
optim = torch.optim.SGD([x], lr=0.001)
print(x)        # tensor([1., 2.], requires_grad=True)
optim.step()
print(x)        # tensor([0.8500, 1.8500], requires_grad=True)
```
Changing the order of `loss_1.backward()` and `loss_2.backward()` will give the same result.

**Comment by FeiiYin** [comment]
Thanks for your instant reply for solving my problems!
